### ISSUE TITLE
JavaScript Backend: Ints are considered to be BigInts

### ISSUE THREAD

**Issue by DoctorRyner** [idris-lang/Idris2#801]
# Steps to Reproduce

1. Create a file named "Main.idr"

```idris
module Main

%foreign "javascript:lambda:x => console.log(x)"
prim__log : Int -> PrimIO ()

log : Int -> IO ()
log = primIO . prim__log

main : IO ()
main = log 1
```

2. Compile it with `idris2 Main.idr --codegen javascript -o index.js`
3. Run it with `node build/exec/index.js`

# Expected Behavior

Seeing 1

# Observed Behavior

Seeing 1n

Int in considered as BigInt when it really should correspond to a rounded number, while Integer should stay to be BigInt

**Comment by rbarreiro** [comment]
why?

**Comment by DoctorRyner** [comment]
Because when we use restricted Int we don't really expect to get an unrestricted BigInt. BigInts are in fact rarely used and in most cases we don't want to use them

**Comment by rbarreiro** [comment]
The Number in JavaScript is really a float, hence I am not sure they are a better representation of Int.

**Comment by DoctorRyner** [comment]
Well, PureScript and Elm handle it this way

**Comment by DoctorRyner** [comment]
Oh, and Nat is considered to be a `BigInt` as well, so it means that I can't use Nat for indexing in JS

**Comment by rbarreiro** [comment]
I think that both Elm and PureScript were made before BigInt was widely supported. 
I don't think Nat is supposed to be bounded.

**Comment by DoctorRyner** [comment]
And I think they don't use BigInt so this simple code wouldn't cause any errors

<img width="415" alt="Screenshot 2020-11-29 at 1 40 37 PM" src="https://user-images.githubusercontent.com/24430919/100539576-8aa47580-3248-11eb-80da-a8f9226c4f04.png">
<img width="499" alt="Screenshot 2020-11-29 at 1 40 47 PM" src="https://user-images.githubusercontent.com/24430919/100539578-8d06cf80-3248-11eb-8e7f-51db2c67a5ed.png">

**Comment by DoctorRyner** [comment]
But I'm not sure at all why I get this error

**Comment by stefan-hoeck** [comment]
Sorry for chiming in on this discussion, but I have had to deal with this issue just recently.

Personally, I think the default of using BigInt for representing Integer types is a valid choice. JavaScript using floating point numbers even for indexing into arrays is just plain broken. While it makes sense to look at how other languages handled this issue, I think we should prefer correctness over ease of use in foreign function calls in this case. Personally, I don't like the idea of the additional mental burden about potential rounding errors when performing integer arithmetic.

However, I've been experimenting with some additional utilities and bindings to the JavaScript world. An approach I used there was to define a newtype wrapper `JSInt` around Double, only implementing interfaces `Num`, `Neg` and `Integral`. Such a type could be used directly in foreign function calls where JavaScript "integers" are required. In addition, its name would give a hint about the semantics to expect from a value of that type. @DoctorRyner, in your example code, you could replace `Int` with this `JSInt` newtype and everything should work as expected after you adjusted your foreign function calls accordingly.

**Comment by DoctorRyner** [comment]
It's an annoyance, the most logical decision would be representing **Int** as **number** and **Integer** as **BigInt**

**Comment by DoctorRyner** [comment]
No one needs it, no one uses it, so why would such a popular type default to BigInt that is rarely used and cases bugs? It's a pain to deal with

**Comment by DoctorRyner** [comment]
Using `BigInt` for `Int` is almost the same as using `Symbol` instead of `string`

**Comment by stefan-hoeck** [comment]
Is the workaround with a wrapper around Double such a pain to deal with, when in return you get clear semantics without having to sacrifice the correct behavior of Int and Nat to a backend without a proper native integer type?

Also, please soften your tone a bit. In my opinion, Idris is more about code correctness than about repeating questionable design decisions in existing programming languages. You opinion might be a different one and that's just fine.

**Comment by DoctorRyner** [comment]
To sum up my points, I wrote next points and assembled them in a one place:

1. **No one uses** uses `BigInt`s
2. Hence it's mostly incompatible with 99% most of JS world where `number`s are used, no one actually cares to use `BigInt` almost like ever
3. `BigInt` was created to represent a very big, unrestricted integer values the similar way as Idris's `Integer`
4. Hence using it for restricted `Int` is weird
5. BUGS,BUGS,BUGS,ANNOYANCE,COERCING,WRAPPING,BUGS,BUGS...
6. PureScript, Elm and GHCJS not only don't use `BigInt`s, they don't seem to even care about it like most of us
7. I can assure you that in the feature there will be people like me complaining about `BigInt`s
8. `BigInt`s are significantly slower than numbers
9. This is like contrary to what a developer would expect from a restricted and rounded `Int` type, once again, `Integer`s are fine as `BigInt`s
10. `BigInt`s can't be used for cryptography
11. Once again, you can't mix `BigInt`s and `number`s, that will result in a type error, and more than 99% of JS code is without `BigInt`s, so there is possibility that using Idris's `Int` can cause a bug in 99% of JS code, why risk so much?
12. Just whyyyy?
13. Why care for some possible **ghost** ugly conversion bugs when we have constant ones about mixing `BigInt`s with numbers? And to resolve them we use, aha, these `number`s that can cause ugly conversion bug so nothing really changes here
14. It isn't obvious that `Int` are represented as `BigInt`s
15. It would be obvious if we had a JS specific package called `Data.BigInt`
16. Possible browser support issues, I'm not sure, but I think browser support of `number`s is better than `BigInt`s ( https://caniuse.com/?search=bigint )

17. `BigInt`s are annoying
18. Having `BigInt`s as a default `Int` representation is even more annoying
19. Older versions of node don't support `BigInt`s
20. `BigInt`s are unexpected, no one thinks about them wring code, so it'll lead to developers making errors and getting frustration
21. Reasoning using `BigInt`s as a type that will allow us to avoid causing implicit ugly floating point numbers conversion errors by making developer causing ugly floating point numbers conversion errors explicitly, doesn't really sound right to me
22. Why `Int` and `Integer` have the same representation as `BigInt`?
23. How are they different then?
24. Why should they be different?
25. Why don't just give developers choice to use `number`'s representation or `BigInt`'s, y implementing `Int` and `Integer`differently?
26. This will allow people to avoid a lot of unexpected bugs
27. People won't need to do any weird wrapping solutions of kinds which @stefan-hoeck described
28. Why would you force people to come up with such weird wrapping solutions in the first place?
29. Dealing with `BigInt`s is cumbersome right now
30. Using `BigInt` has no good things about it, it has no meaning, only inconveniences

**Comment by DoctorRyner** [comment]
@stefan-hoeck 

> repeating  questionable design decisions in existing programming languages

From my point of view, using `BigInt`'s is a questionable language design decision, and probably it's also repeating. Practically speaking, this "correct" implementation doesn't change and fix anything, it only breaks code compatibility

**Comment by gallais** [comment]
@DoctorRyner Please try to stick to a constructive mode of interaction.
This huge list is unnecessary: most of the points are devoid of any argument.

The single most important property of whichever representation we pick
as a mapping of `Int` on the JS side is the guarantee that it behaves in
the same manner as the `Int`s on the Idris side. Being able to prove an
equality on one side of the barrier that does not hold on the other can
have catastrophic consequences. If we can achieve that and have better
performance than going via BigInt then that sounds like a proposal / PR
that could get accepted.

**Comment by DoctorRyner** [comment]
Ok, I get it

**Comment by Pzixel** [comment]
Hi everybody. Let me jump into the thread.

> The single most important property of whichever representation we pick as a mapping of Int on the JS side is the guarantee that it behaves in the same manner as the Ints on the Idris side.

AFAIK idris `int` is 32-signed integer. I didn't find any specifications but here are several issues about it ([link](https://github.com/idris-lang/Idris-dev/issues/2842)) and also documentation that implies that `i64`-returning functions may overflow (see Prelude.File.fileSize for instance).

OTOH `number` in js is a `f64` float number which is able to precisely cover `i32` integer number without rounding errors: an IEEE754 64-bit double can represent any 32-bit integer, simply because it has 53-odd(a) bits available for precision and the 32-bit integer only needs, well, 32

Using native `number` type simplifies lots of use cases since it removes need in casts and has significantly better performance (up to 100 times slower according to [some benchmarks](https://medium.com/better-programming/the-downsides-of-bigints-in-javascript-6350fd807d)). 

I think that aforementioned profits significantly justify proposed changes.

**Comment by Russoul** [comment]
IMO, one shouldn't assume that `Int` represents 32-bit signed integer values. Personally, I believe it is overused in Idris: `Int` is a type without guarantees ! 
For the reason that its runtime size is machine (and backend) dependent *and* so is its computational behaviour. We have more precise types, namely `BitsXX`. However, I think, using `Int` for narrow, language specific computations is reasonable for maximum performance and compatibility.

To sum up, IMO, `Int` should be backed up by `Number` on JavaScript backend.
But as @gallais mentioned, that would probably break some of the code that relies on propositional logic on `Int`s.
That said, we should probably cut back on our internal `Int` use that is supposed to work on every backend. And leverage `BitXX` types instead. Until then, we have to state and formalise their semantics though, like how to handle overflows and such.
